evm/src/arithmetic/columns.rs

```diff
@@ -45,45 +49,66 @@ pub(crate) const ALL_OPERATIONS: [usize; 17] = [
+
+// The auxiliary input columns overlap the general input columns
+// because they correspond to the values in the second row for modular
+// operations.
+const AUX_INPUT_0: Range<usize> = START_SHARED_COLS..START_SHARED_COLS + 2 * N_LIMBS;
 const AUX_INPUT_1: Range<usize> = AUX_INPUT_0.end..AUX_INPUT_0.end + 2 * N_LIMBS;
-const AUX_INPUT_2: Range<usize> = AUX_INPUT_1.end..AUX_INPUT_1.end + N_LIMBS;
+// These auxiliary input columns are awkwardly split across two rows,
+// with the first half after the general input columns and the second
+// half after the auxiliary input columns.
+const AUX_INPUT_2_LO: Range<usize> = GENERAL_INPUT_3.end..GENERAL_INPUT_3.end + N_LIMBS / 2;
+const AUX_INPUT_2_HI: Range<usize> = AUX_INPUT_1.end..AUX_INPUT_1.end + N_LIMBS / 2;
 
+// ADD takes 3 * N_LIMBS = 48 columns
 pub(crate) const ADD_INPUT_0: Range<usize> = GENERAL_INPUT_0;
 pub(crate) const ADD_INPUT_1: Range<usize> = GENERAL_INPUT_1;
 pub(crate) const ADD_OUTPUT: Range<usize> = GENERAL_INPUT_2;
 
+// SUB takes 3 * N_LIMBS = 48 columns
 pub(crate) const SUB_INPUT_0: Range<usize> = GENERAL_INPUT_0;
 pub(crate) const SUB_INPUT_1: Range<usize> = GENERAL_INPUT_1;
 pub(crate) const SUB_OUTPUT: Range<usize> = GENERAL_INPUT_2;
 
+// MUL takes 4 * N_LIMBS = 64 columns
 pub(crate) const MUL_INPUT_0: Range<usize> = GENERAL_INPUT_0;
 pub(crate) const MUL_INPUT_1: Range<usize> = GENERAL_INPUT_1;
 pub(crate) const MUL_OUTPUT: Range<usize> = GENERAL_INPUT_2;
 pub(crate) const MUL_AUX_INPUT: Range<usize> = GENERAL_INPUT_3;
 
+// LT and GT take 4 * N_LIMBS = 64 columns
 pub(crate) const CMP_INPUT_0: Range<usize> = GENERAL_INPUT_0;
 pub(crate) const CMP_INPUT_1: Range<usize> = GENERAL_INPUT_1;
 pub(crate) const CMP_OUTPUT: usize = GENERAL_INPUT_2.start;
 pub(crate) const CMP_AUX_INPUT: Range<usize> = GENERAL_INPUT_3;
 
+// MULMOD takes 4 * N_LIMBS + 2 * 2*N_LIMBS + N_LIMBS = 144 columns
+// but split over two rows of 72 columns each.
+//
+// ADDMOD, SUBMOD, MOD and DIV are currently implemented in terms of
+// the general modular code, so they also take 144 columns (also split
+// over two rows).
 pub(crate) const MODULAR_INPUT_0: Range<usize> = GENERAL_INPUT_0;
-pub(crate) const MODULAR_INPUT_1: Range<usize> = GENERAL_INPUT_1;
+pub(crate) const MODULAR_OUT_AUX_RED: Range<usize> = GENERAL_INPUT_1;
 pub(crate) const MODULAR_MODULUS: Range<usize> = GENERAL_INPUT_2;
 pub(crate) const MODULAR_OUTPUT: Range<usize> = GENERAL_INPUT_3;
 pub(crate) const MODULAR_QUO_INPUT: Range<usize> = AUX_INPUT_0;
 // NB: Last value is not used in AUX, it is used in MOD_IS_ZERO
 pub(crate) const MODULAR_AUX_INPUT: Range<usize> = AUX_INPUT_1;
 pub(crate) const MODULAR_MOD_IS_ZERO: usize = AUX_INPUT_1.end - 1;
-pub(crate) const MODULAR_OUT_AUX_RED: Range<usize> = AUX_INPUT_2;
+pub(crate) const MODULAR_INPUT_1_LO: Range<usize> = AUX_INPUT_2_LO;
+pub(crate) const MODULAR_INPUT_1_HI: Range<usize> = AUX_INPUT_2_HI;
```

Comment:
The current CTL code assumes that all the data lives in the same row, and it can’t compare a row in one table against a pair of rows in another table (please correct me if this is wrong, @wborgeaud).

There is an issue here, which is that while most operations need to cross-reference `MODULAR_INPUT_0`, `MODULAR_INPUT_1`, `MODULAR_MODULUS`, and `MODULAR_OUTPUT` against the CPU, the division operation replaces `MODULAR_OUTPUT` with the low half of `MODULAR_QUO_INPUT` (am I understanding this right, @unzvfu?).

That’s a total of 80 limbs, so it won’t fit in one row as it currently stands. Possible solutions:
1. Make the table 80 columns wide (this is probably the easiest, even if not optimal)
2. Split out the code for division so `ADDMOD`, `SUBMOD`, `MULMOD`, and `MOD` can sync `MODULAR_OUTPUT` against the CPU while `DIV` exposes the low part of `MODULAR_QUO_INPUT`.
3. Change the CTL code so we can use `nv` as well as `lv` in the CTL API (@wborgeaud, I don’t know how much work this would be)

Reply:
> The current CTL code assumes that all the data lives in the same row, and it can’t compare a row in one table against a pair of rows in another table (please correct me if this is wrong, @wborgeaud).

That's correct 👍 

> 3. Change the CTL code so we can use `nv` as well as `lv` in the CTL API (@wborgeaud, I don’t know how much work this would be)

This should be doable without too much effort, let me know if you want to go with this and I'll implement it.

Reply:
> There is an issue here, which is that while most operations need to cross-reference `MODULAR_INPUT_0`, `MODULAR_INPUT_1`, `MODULAR_MODULUS`, and `MODULAR_OUTPUT` against the CPU, the division operation replaces `MODULAR_OUTPUT` with the low half of `MODULAR_QUO_INPUT` (am I understanding this right, @unzvfu?).

@nbgl Yep, spot on.

Regarding the potential solution

>  3. Change the CTL code so we can use `nv` as well as `lv` in the CTL API (@wborgeaud, I don’t know how much work this would be)

this seems like it would be a good idea if it will (or might) be useful for other operations, but otherwise probably not worth complicating the interface just support that one operation. `DIV` is the only operation for which this has come up for me; have either of you two seen other places where adding `nv` to the CTL API would help?

Reply:
> have either of you two seen other places where adding nv to the CTL API would help?

I don't think I have.

Reply:
Updated to use 80 columns instead of 72.